MVVM Sturcture Implementation

AJ-iOSDev opened the pull request:
CollectionView Diffable datasource with MVVM Sturcture - Inprogress

Commit:
MVVM Structure & Diffable datasource -  In progress

Commit:
Created ViewModel & Collectionview Datasource

JoeHoHH commented on AppOne/AppOne/Model/Brands.swift:
```diff
@@ -0,0 +1,16 @@
+//
+//  Brands.swift
+//  AppOne
+//
+//  Created by Asapu, Ajaikumar on 20.11.22.
+//
+
+struct Brands : Hashable {
+    let name: String
+    let headQuarters: String
+    let ceo: String
+
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(name)
+    }
```
can be removed, compiler synthesizes it, also the hash only over the name is not very reliable

JoeHoHH replied:
not sure why you deleted this 😆

JoeHoHH commented on AppOne/AppOne/Model/SectionTitles.swift:
```diff
@@ -0,0 +1,14 @@
+struct SectionTitles : Hashable {
+    let title: String
+
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(title)
+    }
```
can be removed

AJ-iOSDev replied:
don't we need hasher ever here?

JoeHoHH replied:
If you conform to Hashable and all properties conform to Hashabale, then you get the default implementation over all properties for free from the compiler

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,39 @@
+
+import UIKit
+
+class HomeVC: UIViewController {
```
final

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,39 @@
+import UIKit
+
+class HomeVC: UIViewController {
+    private var brandsCollectionView: UICollectionView!
+    private var dataSource: UICollectionViewDiffableDataSource<SectionTitles, Brands>!
```
dont use force unwrap

JoeHoHH commented on AppOne/AppOne/Model/SectionTitles.swift:
```diff
@@ -0,0 +1,14 @@
+//  Created by Asapu, Ajaikumar on 21.11.22.
+//
+
+struct SectionTitles : Hashable {
```
singular `SectionTitle`
```suggestion
struct SectionTitle : Hashable {
```

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    func configureCollectionView() {
+        brandsCollectionView = UICollectionView(frame: view.bounds, collectionViewLayout: configureLayout())
```
this should get it"s size over auto layout

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,39 @@
+
+    func configureCollectionView() {
+        brandsCollectionView = UICollectionView(frame: view.bounds, collectionViewLayout: configureLayout())
+        brandsCollectionView.register(UICollectionViewCell.self, forCellWithReuseIdentifier: "Cell")
```
this is not needed there is a `UICollectionView.CellRegistration`. This is the recommended way to go.

AJ-iOSDev replied:
Ok

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,39 @@
+
+    func configureDataSource() {
+        dataSource = UICollectionViewDiffableDataSource(collectionView: brandsCollectionView, cellProvider: { collectionView, indexPath, itemIdentifier in
+            let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath)
```
`collectionView.dequeueConfiguredReusableCell(using: cellReg, for: indexPath, item: itemIdentifier)`
and then set the items content in the cellreg

JoeHoHH commented:
You completly missed to fill the datasource. so you need to have a snapshot and fill it with the section and the data. Have a look at `NSDiffableDataSnapshot`

JoeHoHH commented on AppOne/AppOne/HomeVC.swift:
```diff
@@ -1,15 +0,0 @@
-//
-//  HomeVC.swift
-//  AppOne
-//
-//  Created by Asapu, Ajaikumar on 19.11.22.
-//
-
-import UIKit
-
-class HomeVC: UIViewController {
-    override func viewDidLoad() {
-        super.viewDidLoad()
-        print("HomeVC")
-    }
-}
```
Why is this deleted? somehow it is not the version from main.....

JoeHoHH commented on AppOne/AppOne/Model/Brands.swift:
```diff
@@ -0,0 +1,16 @@
+//  Created by Asapu, Ajaikumar on 20.11.22.
+//
+
+struct Brands : Hashable {
```
```suggestion
struct Brand : Hashable {
```

Commit:
Inprogress

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,52 @@
+    func configureDataSource() {
+        let cellRegistration = UICollectionView.CellRegistration<SectionTitles, Brands> { cell, indexPath, item in
+            var contentConfiguration = cell.defaultContentConfiguration()
+            contentConfiguration.text = "\(item)"
```
here you should use `item.name`

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    func applySnapshot() {
+        var snapshot = NSDiffableDataSourceSnapshot<SectionTitles, Brands>()
```
you still dont fill the snapshot

JoeHoHH replied:
go with  ```
snapshot.appendSections([some Sectiontitle])
        snapshot.appendItems([Brand])
```

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,52 @@
+        applySnapshot()
+    }
+
+    func configureCollectionView() {
+//        brandsCollectionView = UICollectionView(frame: view.bounds, collectionViewLayout: configureLayout())
```
we somehow have to let the collectionView know it's size

Commit:
Refactored as suggested

JoeHoHH commented on AppOne/AppOne/SceneDelegate.swift:
```diff
@@ -16,8 +16,8 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
         // This delegate does not imply the connecting scene or session are new (see `application:configurationForConnectingSceneSession` instead).
-        guard let _ = (scene as? UIWindowScene) else { return }
-        window = UIWindow()
+        guard let windowscene = (scene as? UIWindowScene) else { return }
```
```suggestion
        guard let windowscene = scene as? UIWindowScene else { return }
```

JoeHoHH replied:
```suggestion
        guard let windowScene = (scene as? UIWindowScene) else { return }
```

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,53 @@
+        let collectionV = UICollectionView(
+            frame: view.bounds,
+            collectionViewLayout: UICollectionViewCompositionalLayout.list(using: UICollectionLayoutListConfiguration(appearance: .plain)))
+        return collectionV
```
you miss the `translatesAutoresizingMaskIntoConstraints`

AJ-iOSDev replied:
Ok moved it to here

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,53 @@
+final class HomeVC: UIViewController {
+    private lazy var brandsCollection: UICollectionView = {
+        let collectionV = UICollectionView(
+            frame: view.bounds,
```
should be set to .zero to let Auto Layout do the sizing

AJ-iOSDev replied:
Ok

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,53 @@
+    private lazy var dataSource: UICollectionViewDiffableDataSource<SectionTitles, Brands> = {
+        let cellRegistration = UICollectionView.CellRegistration<UICollectionViewListCell, Brands> { cell, index, res  in
+            var contentConfig = cell.defaultContentConfiguration()
+            contentConfig.text = res.name
```
here you could also set image and /or secondary text if needed

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,53 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.addSubview(brandsCollection)
+        brandsCollection.translatesAutoresizingMaskIntoConstraints = false
```
should move up no need to split. whole config code for ui component would be in one modular place then

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,53 @@
+
+    private func applySnapshot(sec: SectionTitles, bran: Brands) {
+        var snapshot = NSDiffableDataSourceSnapshot<SectionTitles, Brands>()
+        snapshot.appendSections([sec])
```
it is correct here with the current code, but we could make sections an enum and let it conform to `CaseIterable`, but we can discuss this later.

AJ-iOSDev replied:
Ok

JoeHoHH replied:
You can try an enum and conform to `CaseIterable`

JoeHoHH commented:
Good job, just some minor things. Almost there. 😄

Commit:
Updated suggested changes

AJ-iOSDev commented:
tried playing with custom image and more rows in collectionview. next steps would be that and proper data handling by removing static data in viewdidload

AJ-iOSDev commented:
and also having doubt why would we remove hasher

Commit:
fetching actual data.
Doubts: Don't understand how append sections work here.
/// Metatypes such as `Int.Type` implicitly conform to the `Sendable` protocol.
public protocol Sendable {
}

AJ-iOSDev commented:
Doubts: Don't understand how append sections work here.
/// Metatypes such as `Int.Type` implicitly conform to the `Sendable` protocol.
public protocol Sendable {
}

If datasource gets only item data then shouldn't we send data in sections? only int have to send (kind of no.of sections)?

JoeHoHH commented:
I dont understand your question for Sendable? Why is this connected to you?

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,57 @@
+final class HomeVC: UIViewController {
+    private lazy var brandsCollection: UICollectionView = {
+        let collectionV = UICollectionView(
+            frame: view.bounds,
```
still we dont use `.zero` here

JoeHoHH commented on AppOne/AppOne/View/HomeVC.swift:
```diff
@@ -0,0 +1,57 @@
+            var contentConfig = cell.defaultContentConfiguration()
+            contentConfig.text = res.name
+            contentConfig.secondaryText = res.headQuarters
+//            contentConfig.image = UIImage(named: "Flower.png")
```
why did you comment it here? should work. right?

AJ-iOSDev replied:
no not working, only system image working but custom image shows too big

JoeHoHH replied:
that can be changed over imageprovider

Commit:
Merge branch 'main' into MVVM-Structure

JoeHoHH commented:
> Doubts: Don't understand how append sections work here. /// Metatypes such as `Int.Type` implicitly conform to the `Sendable` protocol. public protocol Sendable { }
> 
> If datasource gets only item data then shouldn't we send data in sections? only int have to send (kind of no.of sections)?

What is your doubt? Normally it is straightforward, a Collectionview has Sections + items in that section. Also it is possible to have 1...n sections. Then you just give the datasource the sections and then you apply the data per section. Not sure where your doubt/misunderstanding is?
